File: Swift/백준/Gold/17298. 오큰수/오큰수.swift
import Foundation

struct Stack<T> {
    var arr = [T]()
    
    var isEmpty: Bool {
        return arr.isEmpty
    }
    var top: T {
        return arr[arr.count - 1]
    }
    mutating func push(_ element: T) {
        arr.append(element)
    }
    mutating func pop() -> T? {
        return arr.isEmpty ? nil : arr.popLast()
    }
}


let n = Int(readLine()!)!

let input = readLine()!.split(separator: " ")

var arr = Array(repeating: -1, count: n)
for i in 0..<arr.count {
    arr[i] = Int(input[i])!
}

var stack = Stack<Int>()

for i in 0..<arr.count {
    
    while !stack.isEmpty && arr[stack.top] < arr[i] {
        let pop = stack.pop()!
        arr[pop] = arr[i]
    }
    
    stack.push(i)
}

while !stack.isEmpty {
    let pop = stack.pop()!
    arr[pop] = -1
}
var result = ""
for i in 0..<arr.count {
    result += String(arr[i])
    result += " "
}
print(result)

File: 백준/Gold/17298. 오큰수/오큰수.java
import java.util.*;
import java.io.*;
public class Main {
	static ArrayList<Integer[]> baad = new ArrayList<>();
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder sb = new StringBuilder();
		int n = Integer.parseInt(br.readLine());
		int arr[] =Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
		Stack<Integer> stack = new Stack<>();
		int result[] = new int[n];

		for(int i=0;i<arr.length;i++) {

			while(!stack.isEmpty()) {
				if(arr[stack.peek()]<arr[i]) {
					result[stack.pop()] = arr[i];
				} else {
					break;
				}
			}


			stack.add(i);

		}
		while(!stack.isEmpty()) {
			result[stack.pop()] = -1;
		}

		for(int i=0;i<result.length;i++) {
			sb.append(result[i]).append(' ');
		}

		System.out.print(sb.toString());

	}

}
